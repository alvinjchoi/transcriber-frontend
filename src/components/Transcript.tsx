import * as React from "react"
import ReactGA from "react-ga"
import { connect } from "react-redux"
import { RouteComponentProps } from "react-router"
import { ActionCreators } from "redux-undo"
import { functions } from "../firebaseApp"
import { ITranscript } from "../interfaces"
import { selectTranscript } from "../store/actions/transcriptActions"
import Paragraphs from "./Paragraphs"

interface IProps {
  history: History
  transcriptId: string
}

interface IReduxStateToProps {
  transcript: {
    future: ITranscript[]
    past: ITranscript[]
    present: ITranscript
  }
}

interface IReduxDispatchToProps {
  clearHistory: () => void
  selectTranscript: (transcriptId: string, transcript: ITranscript) => void
}

class Transcript extends React.Component<RouteComponentProps<{}> & IReduxStateToProps & IReduxDispatchToProps> {
  public componentDidMount() {
    const { transcripts, transcriptId } = this.props

    if (transcriptId !== undefined && transcripts !== undefined) {
      const transcript = transcripts[transcriptId]

      this.props.selectTranscript(transcriptId, transcript)
    }
  }

  public componentDidUpdate(prevProps: IReduxStateToProps & IReduxDispatchToProps) {
    const transcriptId = this.props.transcriptId

    // If transcript has not yet been loaded, or user has selected a new ID, select transcript
    if (Object.entries(this.props.transcript.present).length === 0 || transcriptId !== prevProps.transcriptId) {
      const transcript = this.props.transcripts[transcriptId]

      this.props.selectTranscript(transcriptId, transcript)
    }

    // When the paragraphs are loaded the first time, we reset the undo history
    // This is to stop users from undoing back to a state before the paragraphs were loaded
    if (prevProps.transcript.present.paragraphs === undefined && this.props.transcript.present.paragraphs !== undefined) {
      this.props.clearHistory()
    }
  }

  public render() {
    const transcript = this.props.transcript.present

    // Loading from Firebase
    if (Object.entries(transcript).length === 0) {
      return "Laster inn"
    } else {
      return (
        <main id="transcript">
          <section className="org-bar">
            <span className="org-text-l">{transcript.name}</span>

            <button className="org-btn" onClick={() => this.handleExportTranscriptButtonClicked("docx")}>
              <svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg" width="20" height="20">
                <g fill="none" fillRule="evenodd">
                  <path d="M17 0H3a3 3 0 0 0-3 3v14a3 3 0 0 0 3 3h14a3 3 0 0 0 3-3V3a3 3 0 0 0-3-3z" fill="#252627" />
                  <text fontFamily="Roboto-Medium, Roboto" fontSize="15" fontWeight="400" fill="#FFF" transform="translate(0 -2)">
                    <tspan x="4.4" y="16">
                      w
                    </tspan>
                  </text>
                </g>
              </svg>{" "}
              docx
            </button>

            <button className="org-btn" onClick={() => this.handleExportTranscriptButtonClicked("xmp")}>
              <svg width="20" height="20" focusable="false" aria-hidden="true">
                <use xlinkHref="#icon-premiere" />
              </svg>{" "}
              xmp
            </button>

            <button className="org-btn" onClick={this.handleDeleteButtonClicked}>
              <svg width="20" height="20" focusable="false" aria-hidden="true">
                <use xlinkHref="#icon-garbage" />
              </svg>{" "}
              Slett
            </button>
          </section>
          <div className="paragraphs">
            <Paragraphs transcriptId={transcript.id} />
          </div>
        </main>
      )
    }
  }

  private handleExportTranscriptButtonClicked = (type: string) => {
    ReactGA.event({
      action: "export button pressed",
      category: "transcript",
      label: type,
    })

    const id = this.props.transcriptId

    window.location.href = `${process.env.REACT_APP_FIREBASE_HTTP_CLOUD_FUNCTION_URL}/exportTranscript?id=${id}&type=${type}`
  }

  private handleDeleteButtonClicked = async (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    ReactGA.event({
      action: "delete button pressed",
      category: "transcript",
    })

    const transcriptId = this.props.transcriptId
    const deleteTranscript = functions.httpsCallable("deleteTranscript")

    try {
      this.props.history.push("/transcripts/")
      await deleteTranscript({ transcriptId })
    } catch (error) {
      console.error(error)
      ReactGA.exception({
        description: error.message,
        fatal: false,
      })
    }
  }
}

const mapStateToProps = (state: State): IReduxStateToProps => {
  return {
    transcript: state.transcript,
    transcripts: state.firestore.data.transcripts,
  }
}

const mapDispatchToProps = (dispatch: Dispatch): IReduxDispatchToProps => {
  return {
    clearHistory: () => dispatch(ActionCreators.clearHistory()),
    selectTranscript: (transcriptId: string, transcript: ITranscript) => dispatch(selectTranscript(transcriptId, transcript)),
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Transcript)
