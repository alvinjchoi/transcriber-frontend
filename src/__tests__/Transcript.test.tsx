const firebaseMock = require("firebase-mock")
const mockDatabase = new firebaseMock.MockFirebase()
jest.mock("../firebaseApp", () => {
  const mockStorage = new firebaseMock.MockStorage()
  const mockSdk = new firebaseMock.MockFirebaseSdk(
    // use null if your code does not use RTDB
    (path: string) => {
      return path ? mockDatabase.child(path) : mockDatabase
    },
    // use null if your code does not use AUTHENTICATION
    () => {
      return null
    },
    // use null if your code does not use FIRESTORE
    () => {
      return null
    },
    // use null if your code does not use STORAGE
    () => {
      return mockStorage
    },
    // use null if your code does not use MESSAGING
    () => {
      return null
    },
  )

  // return the mock to match the export api
  return { database: mockSdk.database(), storage: mockSdk.storage }
})

import * as React from "react"
import * as TestRenderer from "react-test-renderer"
import { cleanup, render, /*wait, */ waitForElement } from "react-testing-library"
import Transcript from "../components/Transcript"

jest.mock("react-ga")

test("renders correctly", () => {
  const mock: any = jest.fn()
  const match = { params: { id: "" }, isExact: false, path: "", url: "" }
  const testRenderer = TestRenderer.create(<Transcript match={match} location={mock} history={mock} />)
  expect(testRenderer.toJSON()).toMatchSnapshot()
})

test("show loading text", async () => {
  const mock: any = jest.fn()
  const match = { params: { id: "" }, isExact: false, path: "", url: "" }

  const { getByText } = render(<Transcript match={match} location={mock} history={mock} />)

  await waitForElement(() => getByText("Laster"))
})

test("show transcription not found on invalid id", async () => {
  const mock: any = jest.fn()
  const match = { params: { id: "" }, isExact: false, path: "", url: "" }

  const { getByText } = render(<Transcript match={match} location={mock} history={mock} />)

  mockDatabase.ref.child("/transcripts").push() // Pushing empty object, will result in transaction === null
  mockDatabase.ref.flush()

  await waitForElement(() => getByText("Fant ikke transkripsjonen"))
})
/*
test("show transcription error", async () => {
  const mock: any = jest.fn()
  const match = { params: { id: "" }, isExact: false, path: "", url: "" }

  const { getByText } = render(<Transcript match={match} location={mock} history={mock} />)
  const transcript: any = {
    audioFile: { url: "", name: "" },
    progress: {
      status: Status.Analysing,
    },
    //error: { message: "TEST" },
  }
  mockDatabase.ref.child("/transcripts").push(transcript)
  mockDatabase.ref.flush()

  await wait(() => getByText("Analyserer"))
})
*/
// automatically unmount and cleanup DOM after the test is finished.
afterEach(() => {
  cleanup()
})
