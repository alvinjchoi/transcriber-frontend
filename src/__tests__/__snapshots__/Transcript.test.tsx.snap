// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  id="progress"
>
  <div
    className="dropForm"
  >
    <p
      className="loading"
    >
      Laster inn transkripsjon
    </p>
    <div
      className="react-sweet-progress-circle-outer"
      style={
        Object {
          "fontSize": 27.12,
          "height": 132,
          "width": 132,
        }
      }
    >
      <svg
        className="react-sweet-progress-circle"
        viewBox="0 0 100 100"
      >
        <path
          d="M 50,50 m 0,-47
     a 47,47 0 1 1 0,94
     a 47,47 0 1 1 0,-94"
          fillOpacity="0"
          strokeWidth={6}
          style={
            Object {
              "strokeDasharray": "295.3097094374406px 295.3097094374406px",
              "strokeDashoffset": "-0px",
              "transition": "stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s",
            }
          }
        />
        <path
          d="M 50,50 m 0,-47
     a 47,47 0 1 1 0,94
     a 47,47 0 1 1 0,-94"
          fillOpacity="0"
          stroke="#efefef"
          strokeWidth={6}
          style={
            Object {
              "strokeDasharray": "295.3097094374406px 295.3097094374406px",
              "strokeDashoffset": "-0px",
              "transition": "stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s",
            }
          }
        />
      </svg>
      <div
        className="react-sweet-progress-symbol-absolute"
      >
        <div
          className="react-sweet-progress-symbol"
        >
          ‚è≥
        </div>
      </div>
    </div>
  </div>
</main>
`;
